### PR TITLE
Calico Hosted v2.7

### PR DESCRIPTION
**Pull request by ozdanborne**

### CODE REVIEW

File: playbooks/byo/calico/legacy_upgrade.yml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  tasks:
+  - name: Stop legacy service
+    become: yes
+    systemd:
+      name: calico
+      state: stopped
+  - name: Apply node label
+    delegate_to: "{{ groups.oo_first_master.0 }}"
+    command: >
+      {{ openshift.common.client_binary }} --config={{ openshift.common.config_base }}/master/admin.kubeconfig label node {{ openshift.node.nodename | lower }} --overwrite calicoUpgradeNode=true
```
Comment (caseydavenport) [inline playbooks/byo/calico/legacy_upgrade.yml]: Probably should use `projectcalico.org/ds-ready: true` for the label

File: playbooks/byo/calico/test.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+---
```
Comment (caseydavenport) [inline playbooks/byo/calico/test.yml]: Remove this guy

File: roles/calico_master/README.md [hunk 3]
```diff
@@ -20,7 +14,22 @@ To install, set the following inventory configuration parameters:
...
+then set the following variables in your inventory.ini:
+
+* `calico_etcd_ca_cert_file=/path/to/etcd-ca.crt`
+* `calico_etcd_cert_file=/path/to/etcd-client.crt`
+* `calico_etcd_key_file=/path/to/etcd-client.key`
+* `calico_etcd_endpoints=https://etcd:2379`
+
+## Upgrading
+
+OpenShift-Ansible installs Calico as a hosted install. Previously, calico ran as a systemd service. Running Calico
```
Comment (caseydavenport) [inline roles/calico_master/README.md]: hosted->self-hosted

calico->Calico

File: roles/calico_master/tasks/main.yml [hunk 4]
```diff
@@ -1,26 +1,34 @@
...
 
-- name: Calico Master | Create temp directory for policy controller definition
+- name: Calico Master | oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:calico-node
+  oc_adm_policy_user:
+    user: system:serviceaccount:kube-system:calico-node
+    resource_kind: scc
+    resource_name: privileged
+    state: present
+
+- name: Calico Master | oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:calico-kube-controllers
```
Comment (caseydavenport) [inline roles/calico_master/tasks/main.yml:11]: Why don't we do this in the upgrade playbook?

File: roles/calico_master/templates/calico.yml.j2 [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+  selector:
+    matchLabels:
+      k8s-app: calico-node
+  template:
+    metadata:
+      labels:
+        k8s-app: calico-node
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+        scheduler.alpha.kubernetes.io/tolerations: |
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2]: Don't use tolerations annotation any more

File: roles/calico_master/templates/calico.yml.j2 [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
+           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
+    spec:
+      {% if l_calico_upgrade|default(false) -%}
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2]: Can we get rid of the conditional and just label all new nodes with the "run calico" label?

File: roles/calico_master/templates/calico.yml.j2 [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+                  name: calico-config
+                  key: etcd_endpoints
+            # Choose the backend to use.
+            - name: CALICO_NETWORKING_BACKEND
+              valueFrom:
+                configMapKeyRef:
+                  name: calico-config
+                  key: calico_backend
+            # Cluster type to identify the deployment type
+            - name: CLUSTER_TYPE
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2:165]: add in "origin"

File: roles/calico_master/templates/calico.yml.j2 [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+                  name: calico-config
+                  key: etcd_cert
+            # Auto-detect the BGP IP address.
+            - name: IP
+              value: ""
+            - name: FELIX_HEALTHENABLED
+              value: "true"
+          securityContext:
+            privileged: true
+          resources:
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2]: Probably should either remove of make configurable.

File: roles/calico_master/templates/calico.yml.j2 [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+            - name: CALICO_DISABLE_FILE_LOGGING
+              value: "true"
+            # Set Felix endpoint to host default action to ACCEPT.
+            - name: FELIX_DEFAULTENDPOINTTOHOSTACTION
+              value: "ACCEPT"
+            # Configure the IP Pool from which Pod IPs will be chosen.
+            - name: CALICO_IPV4POOL_CIDR
+              value: "{{ openshift.master.sdn_cluster_network_cidr }}"
+            - name: CALICO_IPV4POOL_IPIP
+              value: "{{ calico_ipv4pool_ipip }}"
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2:182]: Add in K8S_NODEREF

File: roles/calico_master/templates/calico.yml.j2 [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: calico-kube-controllers
+  namespace: kube-system
+  labels:
+    k8s-app: calico-kube-controllers
+  annotations:
+    scheduler.alpha.kubernetes.io/critical-pod: ''
+    scheduler.alpha.kubernetes.io/tolerations: |
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2]: More tolerations

File: roles/calico_master/templates/calico.yml.j2 [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+              name: etcd-certs
+      volumes:
+        # Mount in the etcd TLS secrets.
+        - name: etcd-certs
+          secret:
+            secretName: calico-etcd-secrets
+
+---
+
+# This manifest deploys the Calico policy controller on Kubernetes.
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2]: Let's see if we can remove this part and just do an `oc delete deployment calico-policy-controller` in the upgrade.

File: playbooks/byo/calico/default_versions.yml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+---
+# These variables should match those in roles/calico_master/defaults/main.yml.
```
Comment (ozdanborne) [inline playbooks/byo/calico/default_versions.yml]: fix this comment

File: roles/calico_master/tasks/main.yml (hunk 4 above)
Comment (ozdanborne) [reply roles/calico_master/tasks/main.yml:11]: added.

File: playbooks/byo/calico/legacy_upgrade.yml [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      state: stopped
+  - name: Apply node label
+    delegate_to: "{{ groups.oo_first_master.0 }}"
+    command: >
+      {{ openshift.common.client_binary }} --config={{ openshift.common.config_base }}/master/admin.kubeconfig label node {{ openshift.node.nodename | lower }} --overwrite projectcalico.org/ds-ready=true
+  - name: Wait for node running
+    uri:
+      url: http://localhost:9099/readiness
+      status_code: 204
+    delay: 3
```
Comment (caseydavenport) [inline playbooks/byo/calico/legacy_upgrade.yml:111]: I think we should give a little more breathing room here - maybe a delay of 3 and 5 or 10 retries?

If it succeeded on the 5th time, we wouldn't fail the upgrade.

File: roles/calico/tasks/main.yml [hunk 14]
```diff
@@ -1,105 +1,16 @@
...
-  file: path=/etc/calico/ state=directory
-
-- name: Calico Node | Set calicoctl.cfg
-  template:
-    src: calicoctl.cfg.j2
-    dest: "/etc/calico/calicoctl.cfg"
+    path: /lib/systemd/system/calico.service
+  register: sym
+- fail:
+    msg: You are running a legacy install of Calico. Please run the calico upgrade playbook to upgrade to a hosted install.
```
Comment (caseydavenport) [inline roles/calico/tasks/main.yml]: nit: I suggest changing the wording from "legacy" installation to something like "You are running a systemd based installation of Calico" I think that is friendlier.

File: roles/calico_master/templates/calico.yml.j2 [hunk 15]
```diff
@@ -0,0 +1,400 @@
+---
+kind: ClusterRole
+apiVersion: v1
+metadata:
+  name: calico-kube-controllers
+  namespace: kube-system
+rules:
+  - apiGroups:
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2:8]: We should probably fix this in our upstream manifests - we're giving more permissions than we need here.

e.g. `extensions/pods` isn't required, nor is `""/networkpolicies"`

File: roles/calico_master/templates/calico.yml.j2 [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+      k8s-app: calico-node
+  template:
+    metadata:
+      labels:
+        k8s-app: calico-node
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+    spec:
+      nodeSelector:
+        calicoUpgradeNode: "true"
```
Comment (caseydavenport) [inline roles/calico_master/templates/calico.yml.j2]: We should use `projectcalico.org/ds-ready: true` here.

**Review by caseydavenport — COMMENTED** [review]
@ozdanborne a few more comments but looking good - don't think I need to review again.

**Comment by ozdanborne** [comment]
Submitted upstream. Closing

**Pull request closed without merging**